Add support for ISupportIncrementalLoading interface

marcominerva opened the pull request:
This addition provides helpers class that allow to easily add incremental loading behavior to collections bound to controls like GridView and ListView.

msftclas commented:
Hi __@marcominerva__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

msftclas commented:
__@marcominerva__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSBOT;

ScottIsAFool commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/Incremental Loading Collection/IncrementalLoadingCollectionPage.xaml:
```diff
@@ -0,0 +1,29 @@
+            <RowDefinition Height="*" />
+        </Grid.RowDefinitions>
+        <TextBlock Margin="24" Text="Items are loaded incrementally when the view needs to show them (i.e., when the user scrolls the ListView)" />
+        <Grid Grid.Row="1" Margin="24" BorderThickness="1" BorderBrush="#FFCCCCCC" Background="{StaticResource Brush-Grey-04}" CornerRadius="4">
```
Each property should be on a new line.

ScottIsAFool commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/Incremental Loading Collection/IncrementalLoadingCollectionPage.xaml.cs:
```diff
@@ -0,0 +1,33 @@
+
+namespace Microsoft.Toolkit.Uwp.SampleApp.SamplePages
+{
+    public sealed partial class IncrementalLoadingCollectionPage
```
Missing xml docs

ScottIsAFool commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/Incremental Loading Collection/IncrementalLoadingCollectionPage.xaml.cs:
```diff
@@ -0,0 +1,33 @@
+{
+    public sealed partial class IncrementalLoadingCollectionPage
+    {
+        public IncrementalLoadingCollectionPage()
```
Missing xml docs

ScottIsAFool commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/Incremental Loading Collection/PeopleSource.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace Microsoft.Toolkit.Uwp.SampleApp.SamplePages
+{
+    public class PeopleSource : IIncrementalSource<Person>
```
Missing xml docs

ScottIsAFool commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/Incremental Loading Collection/PeopleSource.cs:
```diff
@@ -0,0 +1,40 @@
+{
+    public class PeopleSource : IIncrementalSource<Person>
+    {
+        private readonly List<Person> people;
```
Private fields should begin with an underscore

ScottIsAFool commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/Incremental Loading Collection/Person.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace Microsoft.Toolkit.Uwp.SampleApp.SamplePages
+{
+    public class Person
```
Missing xml docs on class, public methods/properties

ScottIsAFool commented on Microsoft.Toolkit.Uwp/IncrementalLoadingCollection/IncrementalLoadingCollection.cs:
```diff
@@ -0,0 +1,103 @@
+        /// </param>
+        public IncrementalLoadingCollection(int itemsPerPage = 20)
+        {
+            this.source = new TSource();
```
Remove usages of "this."

ScottIsAFool commented on Microsoft.Toolkit.Uwp/IncrementalLoadingCollection/IncrementalLoadingCollection.cs:
```diff
@@ -0,0 +1,103 @@
+         ISupportIncrementalLoading
+         where TSource : IIncrementalSource<IType>, new()
+    {
+        private readonly TSource source;
```
Private fields should begin with an underscore

ScottIsAFool commented:
Some thoughts on this:
1. Should we include a way to indicate that data is loading? An IsLoading property something that can be bound to?
2. Should we include a way to cancel a request in progress (thinking like #273 has)?
3. As noted, #273 contains an IncrementalLoading implementation, although it only resides in the SampleApp project. Should we take the best bits from both these implementations and move it out of the sample app, as this one is, and then correct the usage in #273?
@deltakosh

deltakosh commented:
1. yes definitely
2. agree
3. we need to have only one version. Let's merge both, move it to Toolkit.UI and correct #273

dotMorten commented:
Looking at this, it seems to me we're just replacing a WinRT interface with a Toolkit one. It might be slightly simpler, but really isn't that much simpler. Is it really worth having this is in a generic toolkit, considering the small value add?

deltakosh commented:
Let's merge #273 first and see what we can do next

marcominerva commented:
@ScottIsAFool @deltakosh about IsLoading, do you mean to create a Dependency Property? Or just to use the ObservableCollection.OnPropertyChanged method when the property actually changes (as it will be read-only)?

ScottIsAFool commented:
@marcominerva I was thinking a dependency property so it can be bound to {Binding Collection.IsLoading} (for example)

marcominerva commented:
@ScottIsAFool IncrementalLoadingCollection already inherits from ObservableCollection. So I need to refactor the code a bit in order to make IncrementalLoadingCollection inherit from DependencyObject (so we can define a Dependency Property on it). Makes sense? :-)

ScottIsAFool commented:
@marcominerva ah yeah, good point.

marcominerva commented:
@ScottIsAFool unfortunately, IncrementalLoadingCollection must inherits from ObservableCollection in order to properly work with incremental loading. I have added a read-only property IsLoading and used the OnPropertyChanged method, tell me if it could be a solution (I have also updated the example to show how it works). In the current PR you'll find also the support for CancellationToken.

dotMorten commented:
> unfortunately, IncrementalLoadingCollection must inherits from ObservableCollection 

Not entirely correct. It just needs to implement `INotifyCollectionChanged`

marcominerva commented:
Sure... But do you mean that it is better to manually implement the INotifyCollectionChanged interface so that IncrementalLoadingCollection can inherits from DependencyObject, rather than to use my simpler solution? According to me it fullfills the requirements, since the IsLoading property is read-only (and the demo shows that it works correctly with databind), but if you think that is better the "DependencyProperty approach", I'll update the code with no-problem :-)

dotMorten commented:
What I'm saying is don't put restrictions on the type of data you can support.
I still don't get the value add of this PR.

There's the WinRT interface:
```
public interface ISupportIncrementalLoading
{
 bool HasMoreItems {get;}
 IAsyncOperation<LoadMoreItemsResult> LoadMoreItemsAsync(uint count);
}
```
Here's yours:
```
public interface IIncrementalSource<TSource>
{
    Task<IEnumerable<TSource>> GetPagedItemsAsync(int pageIndex, int pageSize, CancellationToken? cancellationToken = null);
    bool HasMoreItems { get; }
}
```

They are practically identical - well apart from that yours is more complex, and inadvertently allows you to skip items. The WinRT interface doesn't support skipping, but yours do. I know that's not how you're supposed to use it, but the API allows for it, and as such should handle it.

ScottIsAFool commented:
@dotmorten so what would you suggest that also allows for passing in a CancellationToken?

ScottIsAFool commented on Microsoft.Toolkit.Uwp/IncrementalLoadingCollection/IIncrementalSource.cs:
```diff
@@ -0,0 +1,47 @@
+        /// <returns>
+        /// Returns a collection of <typeparamref name="TSource"/>.
+        /// </returns>
+        Task<IEnumerable<TSource>> GetPagedItemsAsync(int pageIndex, int pageSize, CancellationToken? cancellationToken = null);
```
This should really be:

CancellationToken cancellationToken = default(CancellationToken) 

rather than null.

ScottIsAFool commented on Microsoft.Toolkit.Uwp/IncrementalLoadingCollection/IncrementalLoadingCollection.cs:
```diff
@@ -0,0 +1,162 @@
+        {
+            get
+            {
+                if (_cancellationToken.IsCancellationRequested)
```
If _cancellationToken is null (which it currently would be by default), this would crash the app.

marcominerva replied:
You are absolutely right. Corrected with default(CancellationToken).

marcominerva commented:
>Also don't put a DependencyProperty on a non-control type. If you want to support binding on a model object, implement INotifyPropertyChanged instead.

@dotMorten this is exactly my solution. As I said, I use the `OnPropertyChanged` method of `ObservableCollection` (that already implements `INotifyPropertyChanged`) so you seem to agree with my approach :-)

dotMorten commented:
@ScottIsAFool What do you mean about CancellationToken? Since this is a wrapper for ISupportIncrementalLoading, what will happen when you cancel? You're not really supposed to cancel, since it's not part of the interface being wrapped. 

Besides you can easily add cancellation into your implementation of ISupportIncrementalLoading if you really want it. I still don't see this as a significant value-add to the toolkit. It's more bloat to maintain, and doesn't significantly reduce the work a developer will need to do to get incrementally loading collections.

deltakosh commented:
@dotMorten do you talk about all this PR or just a specific part of it?

We also have this file: https://github.com/Microsoft/UWPCommunityToolkit/pull/273/files#diff-0fabab6a8f6cd88df4fb052192c03781

Which seems comparable

marcominerva commented:
@ScottIsAFool @dotMorten the last version of this PR adds also the IsLoading property that is automatically managed and can be data-bound. Moreover, having parameters like `pageIndex` and `pageSize`makes makes easier to handle pagination. If you manually implement the ISupportIncrementalLoading interface, you always need to write repetitive code to keep track of current page and to get notified when new items are actually loading.

Moreover, you say:

> and inadvertently allows you to skip items. The WinRT interface doesn't support skipping

In which way does my implementation allow to skip items, while the WinRT interface doesn't?

deltakosh commented:
So we integrated this:
https://github.com/Microsoft/UWPCommunityToolkit/blob/dev/Microsoft.Toolkit.Uwp.SampleApp/SamplePages/MicrosoftGraph%20Service/IncrementalCollection.cs

It could be great to remove it and replace it by your implementation.

deltakosh commented:
Hello, we plan to close v1.1 this monday (9/26). There is no rush but if you want to be part of v1.1 we ahve to merge your code before this Friday

marcominerva commented:
So if I replace the IncrementalCollection.cs in Microsoft Graph Service with my implementation, then will you merge my PR?

deltakosh commented:
Actually I would like to see live in this real example. I like the idea of having just a simple interface to provide to support incremental. Let's see how it fits in the graph sample

marcominerva commented:
Ok, I'll update the PR as soon as possible!

marcominerva commented:
Hi @deltakosh! I have update my PR as stated, let me know what do you think about my implementation.

deltakosh commented:
I like it!! It is worth merging to me!

deltakosh commented:
Let's just wait for another reviewer

ScottIsAFool commented on Microsoft.Toolkit.Uwp/IncrementalLoadingCollection/IncrementalLoadingCollection.cs:
```diff
@@ -0,0 +1,207 @@
+                    }
+                    catch (Exception ex) when (_onError != null)
+                    {
+                        _onError.Invoke(ex);
```
_onError?.Invoke(ex) as _onError could be null.

marcominerva replied:
There is an Exception filter on line 167 to check whether _onError is != null, so there is no need to use the ?. operator.

ScottIsAFool replied:
Ah, sorry, missed that.

ScottIsAFool commented on Microsoft.Toolkit.Uwp/IncrementalLoadingCollection/IncrementalLoadingCollection.cs:
```diff
@@ -0,0 +1,207 @@
+                        _onError.Invoke(ex);
+                    }
+
+                    if (data != null && data.Count() > 0 && !_cancellationToken.IsCancellationRequested)
```
Use data.Any() here instead of data.Count() > 0

marcominerva replied:
Done :-)

ScottIsAFool commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/MicrosoftGraph Service/MicrosoftGraphSource.cs:
```diff
@@ -0,0 +1,52 @@
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Text;
```
Remove redundant usings

marcominerva replied:
Done :-)

marcominerva commented:
@ScottIsAFool I have removed the HasMoreItems property from IIncrementalSource to simply the implementation of the interface (in this way it is necessary only a single method). Moreover, now the HasMoreItems property of IncrementalLoadingCollection is observable (in fact I have updated the sample accordingly).

deltakosh commented:
Congrats! now it's time for documentation :)

Do you mind adding a page here: https://github.com/Microsoft/UWPCommunityToolkit-docs

marcominerva commented:
For sure :-) I'll make a new PR with the documentation as soon as possible :-)

deltakosh commented:
Thanks!
